File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as s]))

(def question-mark \?)

(defn is-silence [prompt]
  (empty? (s/trim prompt)))

(defn is-question [prompt]
  (= (last (s/trim prompt)) question-mark))

(defn is-yelling [prompt]
  (and (= prompt (s/upper-case prompt))
       (re-find #"[a-zA-Z]" prompt)))

(defn is-yelling-question [prompt]
  (and (is-yelling prompt) (is-question prompt)))

(defn response-for [prompt]
  (cond
    (is-silence prompt) "Fine. Be that way!"
    (is-yelling-question prompt) "Calm down, I know what I'm doing!"
    (is-question prompt) "Sure."
    (is-yelling prompt) "Whoa, chill out!"
    :else "Whatever."))

File: python/bob/bob.py
def _is_silence(s: str) -> bool:
    return not s.strip()


def _is_question(s: str) -> bool:
    return s.strip()[-1] == "?"


def _is_yelling(s: str) -> bool:
    return s.isupper()


def _is_yelling_question(s: str) -> bool:
    return _is_yelling(s) and _is_question(s)


def response(hey_bob: str) -> str:
    if _is_silence(hey_bob):
        return "Fine. Be that way!"
    elif _is_yelling_question(hey_bob):
        return "Calm down, I know what I'm doing!"
    elif _is_question(hey_bob):
        return "Sure."
    elif _is_yelling(hey_bob):
        return "Whoa, chill out!"
    return "Whatever."
